src/os/bluestore/BitmapFreelistManager.cc

```diff
@@ -0,0 +1,399 @@
+  while (start < bits) {
+    int byte = start >> 3;
+    unsigned char mask = 1 << (7 - (start & 7));
+    if (p[byte] & mask) {
+      return start;
+    }
+    ++start;
+  }
+  return -1; // not found
+}
+
+bool BitmapFreelistManager::enumerate_next(uint64_t *offset, uint64_t *length)
+{
+  std::lock_guard<std::mutex> l(lock);
+
+  // initial base case is a bit awkward
+  if (enumerate_offset == 0) {
+    dout(10) << __func__ << " start" << dendl;
+    enumerate_p = kvdb->get_iterator(prefix);
+    enumerate_p->lower_bound(string("o"));
+    // we assert that the first block is always allocated; it's true,
+    // and it simplifies our lives a bit.
+    assert(enumerate_p->valid());
+    string k = enumerate_p->key();
+    const char *p = k.c_str() + 1;
+    _key_decode_u64(p, &enumerate_offset);
+    enumerate_bl = enumerate_p->value();
+    enumerate_bl_pos = 0;
+    assert(enumerate_offset == 0);
+    assert(get_next_set_bit(enumerate_bl, 0) == 0);
+  }
+
+  // skip set bits to find offset
+  while (true) {
+    enumerate_bl_pos = get_next_clear_bit(enumerate_bl, enumerate_bl_pos);
+    if (enumerate_bl_pos >= 0) {
+      dout(30) << __func__ << " found clear bit, key " << enumerate_offset
+	       << " bit " << enumerate_bl_pos
+	       << " offset " << get_offset(enumerate_offset, enumerate_bl_pos)
+	       << dendl;
+      break;
+    }
+    if (enumerate_bl_pos < 0) {
+      dout(30) << " no more clear bits in " << enumerate_offset << dendl;
+      enumerate_p->next();
+      enumerate_bl.clear();
+      if (!enumerate_p->valid()) {
+	enumerate_offset += bytes_per_block * blocks_per_key;
+	enumerate_bl_pos = 0;
+	break;
+      }
+      string k = enumerate_p->key();
+      const char *p = k.c_str() + 1;
+      _key_decode_u64(p, &enumerate_offset);
+      enumerate_bl = enumerate_p->value();
+      enumerate_bl_pos = 0;
+    }
+  }
+  *offset = get_offset(enumerate_offset, enumerate_bl_pos);
+
+  // skip clear bits to find the end
+  uint64_t end = 0;
+  if (enumerate_p->valid()) {
+    while (true) {
+      enumerate_bl_pos = get_next_set_bit(enumerate_bl, enumerate_bl_pos);
+      if (enumerate_bl_pos >= 0) {
+	end = get_offset(enumerate_offset, enumerate_bl_pos);
+	dout(30) << __func__ << " found set bit, key " << enumerate_offset
+		 << " bit " << enumerate_bl_pos
+		 << " offset " << end
+		 << dendl;
+	*length = end - *offset;
+	dout(10) << __func__ << " " << *offset << "~" << *length << dendl;
+	return true;
+      }
+      if (enumerate_bl_pos < 0) {
+	dout(30) << " no more set bits in " << enumerate_offset << dendl;
+	enumerate_p->next();
+	enumerate_bl.clear();
+	enumerate_bl_pos = 0;
+	if (!enumerate_p->valid()) {
+	  break;
+	}
+	string k = enumerate_p->key();
+	const char *p = k.c_str() + 1;
+	_key_decode_u64(p, &enumerate_offset);
+	enumerate_bl = enumerate_p->value();
+      }
+    }
+  }
+
+  end = blocks * bytes_per_block;
+  if (enumerate_offset < end) {
+    *length = end - enumerate_offset;
```

Comment:
Perhaps we shall introduce a member to remember the `size` passed in during initialization.
```c++
int BitmapFreelistManager::create(uint64_t size, KeyValueDB::Transaction txn)
{
  bytes_per_block = g_conf->bdev_block_size;
  blocks = size / bytes_per_block;
  blocks_per_key = g_conf->bluestore_freelist_blocks_per_key;
  if (blocks / blocks_per_key * blocks_per_key != blocks) {
    dout(10) << __func__ << " rounding blocks up from " << blocks << dendl;
    blocks = (blocks / blocks_per_key + 1) * blocks_per_key;
  }
```
As we may round up `blocks` to be `blocks_per_key` aligned as above, which means here `end` could pass over the origin `size` and cause `offset + length` to be greater than origin `size`.